remote/main.go

```diff
@@ -0,0 +1,142 @@
+type WebsocketTargetJson struct {
+	WebSocketDebuggerUrl string `json:"webSocketDebuggerUrl"`
+	Url                  string `json:"url"`
+}
+
+// replaceResolvedHostnameIfNeeded tries to resolve the hostname in the given URL in order to work around mitigations
+// introduced because of https://bugs.chromium.org/p/chromium/issues/detail?id=813540
+//
+// Chromium will return the error "Host header is specified and is not an IP address or localhost" if the
+// DevTools are connected to via a hostname other than localhost.
+func replaceResolvedHostnameIfNeeded(url *url.URL) error {
+	if url.Hostname() == "localhost" {
+		return nil
+	}
+
+	addresses, err := net.LookupHost(url.Hostname())
+
+	if err != nil {
+		return err
+	}
+
+	if len(addresses) != 1 {
+		log.Printf("Warning: Host lookup of %q returned %v, picking first address\n", url.Hostname(), addresses)
+	}
+
+	port := url.Port()
+	url.Host = fmt.Sprintf("%s:%s", addresses[0], port)
+
+	log.Printf("Replaced hostname with %q\n", addresses[0])
+
+	return nil
+}
+
+// getWsUrl gets the websocket target URL from the default tab in a running chromium instance
+func getWsUrl(chromiumRemoteUrl *url.URL) (string, error) {
+	err := replaceResolvedHostnameIfNeeded(chromiumRemoteUrl)
+
+	if err != nil {
+		return "", err
+	}
+
+	resp, err := http.Get(chromiumRemoteUrl.String())
+
+	if err != nil {
+		return "", err
+	}
+
+	bytes, err := ioutil.ReadAll(resp.Body)
+
+	if err != nil {
+		return "", err
+	}
+
+	err = resp.Body.Close()
+
+	if err != nil {
+		return "", err
+	}
+
+	targets := make([]WebsocketTargetJson, 0)
+	err = json.Unmarshal(bytes, &targets)
+
+	if err != nil {
+		return "", fmt.Errorf("could not parse %q as JSON, error was: %v", string(bytes), err)
+	}
+
+	for _, target := range targets {
+		if target.Url == "about:blank" { // we assume this is the default page and that it is always around
+			return target.WebSocketDebuggerUrl, nil
+		}
+	}
+
+	return "", fmt.Errorf("no default page found on %q", chromiumRemoteUrl.String())
+}
+
+// getHtml opens a URL in a new tab and returns its DOM
+func getHtml(parent context.Context, url string) (string, error) {
+	ctxt, cancel := chromedp.NewContext(parent) // create new tab
+	defer cancel()                              // close tab afterwards
+
+	var body string
+	if err := chromedp.Run(ctxt,
+		chromedp.Navigate(url),
+		chromedp.Sleep(time.Second), // wait for things to settle
+		chromedp.OuterHTML("html", &body),
+	); err != nil {
+		return body, err
+	}
+
+	return body, nil
+}
+
+func main() {
+	var rawChromiumUrl string
+	flag.StringVar(&rawChromiumUrl, "chromium-host", "http://localhost:9222/json",
+		"Pass URL where chromium instance is running.")
+	flag.Parse()
+
+	chromiumUrl, err := url.Parse(rawChromiumUrl)
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	wsUrl, err := getWsUrl(chromiumUrl)
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ctxt, cancel := chromedp.NewRemoteAllocator(context.Background(), wsUrl)
```

Comment:
we generally use `actx` for allocator contexts, and `ctx` for regular contexts.

it would also be better to create both contexts in the same block.

Reply:
Fixed